ambry-tools/src/main/java/com/github/ambry/clustermap/HelixFullAutoReconstructResourceTool.java

```diff
@@ -0,0 +1,247 @@
+      int cliqueId = entry.getKey();
+      List<String> hosts = entry.getValue();
+      cliqueToPartitions.put(cliqueId, new HashSet<>());
+      Set<String> partitionSet = cliqueToPartitions.get(cliqueId);
+      for (String host : hosts) {
+        String instanceName = ClusterMapUtils.getInstanceName(host, port);
+        if (currentStates.containsKey(instanceName)) {
+          List<String> partitions = currentStates.get(instanceName);
+          partitionSet.addAll(partitions);
+        } else {
+          System.err.println("Instance " + instanceName + " is not present in cluster or has empty current state");
+        }
+      }
+    }
+
+    for (Map.Entry<Integer, Set<String>> entry : cliqueToPartitions.entrySet()) {
+      System.out.println("Clique = " + entry.getKey() + ", Partitions = " + entry.getValue());
+    }
+  }
+
+  public void createNewResources() {
+    // 1. For each newly formed clique, add a resource in helix cluster
+    for (Map.Entry<Integer, Set<String>> entry : cliqueToPartitions.entrySet()) {
+      int cliqueId = entry.getKey();
+      Set<String> partitions = entry.getValue();
+      String resource = String.valueOf(cliqueId);
+      Map<String, List<String>> resourcePreferenceLists = new HashMap<>();
+      for (String partition : partitions) {
+        List<String> preferenceList = preferenceLists.get(partition);
+        resourcePreferenceLists.put(partition, preferenceList);
+      }
+      buildAndCreateIdealState(resource, resourcePreferenceLists);
+    }
+  }
+
+  /**
+   * Build and create a new IdealState for the given resource name and partition layout.
+   *
+   * @param resourceName    The name of the newly created resource
+   * @param preferenceLists The partition layout
+   */
+  private void buildAndCreateIdealState(String resourceName, Map<String, List<String>> preferenceLists) {
+    IdealState idealState = buildIdealState(resourceName, preferenceLists);
+    if (!dryRun) {
+      helixAdmin.addResource(clusterName, resourceName, idealState);
+      System.out.println(
+          "Added " + preferenceLists.size() + " new partitions under resource " + resourceName + " in dc " + dc);
+    } else {
+      System.out.println(
+          "Under DryRun mode, " + preferenceLists.size() + " new partitions are added to resource " + resourceName
+              + " in dc " + dc);
+    }
+  }
+
+  /**
+   * Build a new IdealState for the given resource name and partition layout
+   *
+   * @param resourceName    The name of the newly created resource
+   * @param preferenceLists The partition layout
+   * @return The new IdealState
+   */
+  private IdealState buildIdealState(String resourceName, Map<String, List<String>> preferenceLists) {
+    IdealState idealState = new IdealState(resourceName);
+    idealState.setStateModelDefRef(ClusterMapConfig.DEFAULT_STATE_MODEL_DEF);
+    for (Map.Entry<String, List<String>> entry : preferenceLists.entrySet()) {
```

Comment:
let's explicitly set rebalance mode to semi auto, and the ideal state mode to auto, for now